front: fix electrification modes and profiles in csv export

Synar opened the pull request:
Close #10491

(Including the issue pointed out by @Castavo in the comment)

Synar committed:
front: fix electr ranges pos time and speed in csv export

In exportTrainCSV.ts, electrificationStart was already in meters.

Thus stop converting for milimeters to meters when meters are expected,
and convert from meters to milimeters when milimeters are expected.

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: fix electrification ranges and profiles merging

The function formatElectrificationRanges aimed to merge together electrification ranges
and electrical profile ranges.

However, this was done assuming that the boundaries of electrical profiles would be the
exact same or included in the boundaries of electrification ranges, which is not true.

The electrical profiles data produced by formatElectrificationRanges was thus no longer used
in the simulation csvs, as it was incomplete.

In order to use them again, properly merge these ranges together.

The easiest way to properly do so is actually to merge them before converting from boundaries to ranges.

Thus invert the order of these operations.

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: fix electr mode and profile coherence in csv export

In the exported simulation csv, we were adding a new line
each time there was a change in electrical profile,
and each time there was a change in electrification.

Since both often happen at the same time,
there were not only a lot of unnecessary line pairs,
but the first line of each pair had partially outdated information.

Fix this by using already merged electrical profiles and electrification data
thanks to mergeElectrificationAndProfiles.

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Yohh commented on front/src/applications/operationalStudies/__tests__/utils.spec.ts:
```diff
@@ -74,16 +74,16 @@ describe('transformBoundariesDataToRangesData', () => {
 
-  it('should properly format electrification ranges if electrification is longer than electrical profiles', () => {
-    const result = formatElectrificationRanges(
+  it('should properly marge electrification ranges and electrical profiles if they ranges are mismatched', () => {
```
```suggestion
  it('should properly merge electrification ranges and electrical profiles if these ranges are mismatched', () => {
```

Synar replied:
Thanks! I must have been tired when I wrote this haha

Yohh commented on front/src/applications/operationalStudies/utils.ts:
```diff
@@ -113,64 +113,73 @@ export const transformBoundariesDataToRangesData = <
+  range: { start: number; stop: number },
+  segmentStart: number,
+  segmentEnd: number
+) => range && (range.start <= segmentStart || isNaN(range.start)) && range.stop >= segmentEnd;
```
```suggestion
) => (range.start <= segmentStart || isNaN(range.start)) && range.stop >= segmentEnd;
```

Synar replied:
Good catch, thanks!

Synar replied:
Actually this was a typing problem. We are not guaranteed that `sortedElectrifications[electrificationIndex]` is defined, so I'm rewriting it as `sortedElectrifications.at(electrificationIndex)` and letting the function take undefined.

Yohh replied:
ok nice, one way or the other I don't mind

Yohh commented on front/src/applications/operationalStudies/utils.ts:
```diff
@@ -113,64 +113,73 @@ export const transformBoundariesDataToRangesData = <
+        .flatMap((range) => [range.start, range.stop])
+        .concat(sortedProfiles.flatMap((range) => [range.start, range.stop]))
+    )
+  ).sort((a, b) => a - b);
```
```suggestion
   const boundaries = sortedElectrifications
    .flatMap((range) => [range.start, range.stop])
    .concat(sortedProfiles.flatMap((range) => [range.start, range.stop]))
    .sort((a, b) => a - b);
```


not sure about that, but do you really need to set?

Synar replied:
Good catch! Since we handle 0 zero length segments anyway, it should be fine to have some redundant boundaries here.

kmer2016 commented on front/src/applications/operationalStudies/utils.ts:
```diff
@@ -113,64 +113,73 @@ export const transformBoundariesDataToRangesData = <
-        stop: currentElectrification.stop,
-      };
+): ElectrificationRange[] => {
+  const sortedElectrifications = [...electrifications];
+  sortedElectrifications.sort((e1, e2) => e1.start - e2.start);
+  const sortedProfiles = [...electricalProfiles];
+  sortedProfiles.sort((e1, e2) => e1.start - e2.start);
```
```typescript
const sortedElectrifications = [...electrifications].sort((e1, e2) => e1.start - e2.start);
const sortedProfiles = [...electricalProfiles].sort((e1, e2) => e1.start - e2.start);
```

Synar replied:
Sure thanks!

I did this, but then I ended up rewriting this part, so this no longer applies

Synar committed:
front: fix NaN values in electrification and profile ranges

When provided an empty boundary list,
transformBoundariesDataToRangesData added a segment from NaN to the end of the path
instead of from 0 to the end of the path.

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar commented:
I ended up rewriting the middle of the pr, with the middle commit being entirely rewritten from scratch, and a commit to fix the NaN issue was added. 

Not using fixup, as it's a complete rewrite.

kmer2016 reviewed (approved):
LGTM
